src/fparser/two/Fortran2008.py

```diff
@@ -108,6 +116,151 @@ class Program_Unit(Program_Unit_2003):  # R202
+
+    '''
+    subclass_names = [
+        'Action_Stmt', 'Associate_Construct', 'Case_Construct',
+        'Do_Construct', 'Forall_Construct', 'If_Construct',
+        'Select_Type_Construct', 'Where_Construct']
+
+
+class Executable_Construct_C201(Executable_Construct_C201_2003):
+    '''
+    executable-construct-c201 is executable construct.
+    This applies C201.
+    '''
+    subclass_names = Executable_Construct.subclass_names[:]
+    subclass_names[subclass_names.index('Action_Stmt')] = 'Action_Stmt_C201'
+
+
+class Action_Stmt(Action_Stmt_2003):  # R214
+    '''
+    Fortran 2008 rule R214
+    action-stmt is allocate-stmt
+                    or assignment-stmt
+                    or backspace-stmt
+                    or call-stmt
+                    or close-stmt
+                    or continue-stmt
+                    or cycle-stmt
+                    or deallocate-stmt
+                    or end-function-stmt
+                    or end-mp-subprogram-stmt
+                    or end-program-stmt
+                    or end-subroutine-stmt
+                    or endfile-stmt
+                    or error-stop-stmt
+                    or exit-stmt
+                    or flush-stmt
+                    or forall-stmt
+                    or goto-stmt
+                    or if-stmt
+                    or inquire-stmt
+                    or lock-stmt
+                    or nullify-stmt
+                    or open-stmt
+                    or pointer-assignment-stmt
+                    or print-stmt
+                    or read-stmt
+                    or return-stmt
+                    or rewind-stmt
+                    or stop-stmt
+                    or sync-all-stmt
+                    or sync-images-stmt
+                    or sync-memory-stmt
+                    or unlock-stmt
+                    or wait-stmt
+                    or where-stmt
+                    or write-stmt
+                    or arithmetic-if-stmt
+                    or computed-goto-stmt
+
+    The implementation of this rule adds the relevant subclass names
+    for new statements added in Fortran 2008.
+
+    Associated constraints are:
+
+    "C201 (R208) An execution-part shall not contain an end-function-stmt,
+          end-mp-subprogram-stmt, end-program-stmt, or end-subroutine-stmt."
+
+    NB: The following statements are not yet implemented:
+    end-mp-subprogram-stmt, endfile-stmt, lock-stmt, sync-all-stmt,
+    sync-images-stmt, sync-memory-stmt, unlock-stmt.
+
+    '''
+    # Fortran 2008 adds a few additional action-stmt. We therefore
+    # extend the Fortran 2003 specification
+    subclass_names = Action_Stmt_2003.subclass_names[:]
+    subclass_names.append("Error_Stop_Stmt")
+
+
+class Action_Stmt_C201(Action_Stmt_C201_2003):
+    """
+    action-stmt-c201 is action-stmt
+    C201 is applied.
+    """
+    subclass_names = Action_Stmt.subclass_names[:]
```

Comment:
As before, what's the reason for implementing these constraints in a separate sub-class?

Reply:
As before, just carried over the F2003 way of things...